toastr feature/toastr container clone element

Chun-Lin opened the pull request:
## DONE
* Use **cloneElement** to send props data to children
* Use **ReactCSSTransitionGroup** to implement animation feature

There are two ways to see the demos.
1. `yarn styleguide` to see the styles of toastrs
2. `yarn start` to see the demo page

Chun-Lin committed:
feat: add basic toastr style and display it on styleguidist

Chun-Lin committed:
feat: complete basic style of toastr based on react-qui guideline

Chun-Lin committed:
refactor: extract size and animation duration to outside container

Chun-Lin committed:
feat: add width, height, duration props in container

Chun-Lin committed:
feat: add postion props and fix some bugs on width and height props

Chun-Lin committed:
refactor: remove Toastr.propTypes cuz static propTypes existed

Chun-Lin committed:
feat: add shorid modules and reactcsstransitiongroup modules, now the animation can work

Chun-Lin committed:
refactor: move toastrstyle and duration props to toastrContainer and refactor the code

Chun-Lin committed:
refactor: remove useless import

purepennons commented:
@Chun-Lin 
- 有 warning ci 會擋掉
- lib/ 記得進 .gitignore
- ToastrContainer 的 styleguidst markdown 要補使用方法
- 是否要使用 reset.css/normalize.css ?
- 如果 styleguidist 無法修好的話，記得在 readme 上補上用法。

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/Toastr.js:
```diff
@@ -0,0 +1,55 @@
+      'download-st',
+      'video-st',
+    ]),
+    children: PropTypes.oneOfType([
```
children 可以是 array 吧？

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+      })
+    })
+
+    if (this.state.queue.length < 0) return
```
提到 const 之前，可以避免無意義多餘運算。

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  static defaultProps = {
+    duration: 5,
```
當初說統一成毫秒

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+      queue: [...prevState.queue, ...children]
+    }))
+
+    window.setTimeout(() => {
```
養成習慣，要把 return id 存起來，然後在 componentWillUnmount 的時候 clearTimeout，不然可能造成 memory leak。

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+    }
+  }
+
+  trigger = () => {
```
傳入 childProps，改成可以動態修改 children props

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+    width: PropTypes.number,
+    height: PropTypes.number,
+    duration: PropTypes.number,
+    position: PropTypes.string,
```
還缺六個方位的 class。

purepennons commented on packages/toastr/src/react-component-node-modules/index.js:
```diff
@@ -0,0 +1,2 @@
+export { default as Toastr } from './components/Toastr/Toastr';
```
名稱有要統一嗎？邏輯相關就元件名稱 (`<Toastr>`），Q 才能使用的 View 就 `<QToastr>` ?

所以：
- Toastr -> QToastr
- ToastrContainer -> Toastr

Chun-Lin replied:
ok, 我把名稱統一了，邏輯層統一叫做Toastr，有用到QNAP的style就叫做QToastr

purepennons commented:
@Chun-Lin 

如果 styleguidist 無法修好的話，記得在 readme 上補上用法。

Chun-Lin committed:
feat: add dynamic children feature

add a parameter for childProps to let users can modify children dynamically

purepennons commented on packages/toastr/lib/components/Toastr/Toastr.md:
```diff
@@ -0,0 +1,33 @@
+
+const toastrFunc = () =>{
+  console.log('test')
+  toastrContainer.info('Hello!!!', 5)
```
這個 info 是？

Chun-Lin replied:
無意義的程式碼

purepennons commented on packages/toastr/package.json:
```diff
@@ -0,0 +1,43 @@
+    "stylelint": "./node_modules/stylelint/bin/stylelint.js ./src/**/*.{scss,sass}",
+    "test": "react-scripts test --env=jsdom"
+  },
+  "dependencies": {
```
沒用到的 dependencies 可以拔掉。

purepennons replied:
可以的話，幫我試試看拔成 devDependencies ，看會不會壞掉，不然每個套件都太肥了。

Chun-Lin replied:
從Dependencies拔到devDependencies可以work

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/Toastr.js:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  componentDidMount() {
+    this.getContainerStyle(this.props)
```
無意義程式碼，沒 value 去接他

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/Toastr.js:
```diff
@@ -0,0 +1,55 @@
+
+  getContainerStyle = () => {
+    const { width, height } = this.props
+    const widthWithoutPadding = width - 48
```
48 要拉成有意義的變數再使用

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+      return React.cloneElement(child, {
+        width,
+        height,
+        toastrStyle,
```
這是屬於 Toastr 的樣式設定，不應該是 Container 的 props。這是使用時，Toastr 要寫死的。記得一併把 propTypes 移除。

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/ToastrContainer.js:
```diff
@@ -0,0 +1,86 @@
+    }))
+
+    window.setTimeout(() => {
+      let newItems = this.state.queue.slice(1)
```
`newItems ` 並沒有 reassign，請用 const

purepennons replied:
而且這是用於下次變化的 state，在 setState 中的 prevState 設定才安全。

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/styles/animation.css:
```diff
@@ -0,0 +1,17 @@
+.toastranimation-enter {
+  opacity: 0.01;
```
為什麼要先設成 0.01?

Chun-Lin replied:
因為react animation add-on範例是opacity 0.01開始，其實設成0也是可以的，只是因為幾乎沒影響，所以我就沒改成0

purepennons commented on packages/toastr/package.json:
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "react-qui-toastr",
```
名稱需要為 `@react-qui/toastr`

Chun-Lin committed:
fix: remove useless code and refactor code.

Chun-Lin committed:
refactor: rename component and move props

rename Toastr to QToastr, ToastrContainer to Toastr, move those props related to view to from Toastr
to QToastr

Chun-Lin committed:
fix: add 3 more postion and fix some css bugs to display correctly

add a wrapper outside Toastr component and make it a flex container. Also, adding 3 postions, which
are bottom-left, bottom, and bottom-right

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/Toastr.js:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  componentWillUnmount() {
+    if (!this.id) return
```
this.timeID 才對

purepennons commented on packages/toastr/src/react-component-node-modules/components/Toastr/styles/Toastr.scss:
```diff
@@ -0,0 +1,65 @@
+.view-page {
```
給他一個很高的 z-index，不然會被其他東西蓋過去。（甚至設定 opacity: 0.999 也行 XD...）

purepennons commented:
- 另外補上 stylegudist 的 markdown，我試過是可以的
```js
let toastr;
<div>
  <button onClick={evt => toastr.trigger({children: <span>No, changed!</span>})}>trigger</button>
  <Toastr
    position="top-right"
    duration={5000}
    ref={ref => toastr = ref}
  >
    <QToastr toastrStyle="qts">Default Message</QToastr>
  </Toastr>
</div>
```

- center 在某些情況下會偏移，還不確定原因 （不管是 top or bottom）
<img width="1312" alt="screen shot 2018-01-09 at 5 01 30 pm" src="https://user-images.githubusercontent.com/4588113/34712921-8573808e-f55f-11e7-9eff-f44f05e2fe34.png">

Commit:
fix: add z-index and move QToastr component to QToastr folder

add z-index to Toastr.scss and move QToastr and its styles to QToastr folder

Commit:
fix: add zIndex as a Toastr props and deleted lib directory

add zIndex as a Toastr props, users can set zIndex value manually on Taostr component. Also, I
deleted old lib directory.

Commit:
fix: modify duration time in App.js to 5000ms

Commit:
fix: add zIndex props
